### PR TITLE
fix: bugs related to merge scan

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Fix some bugs related distributed query
 - prevent `TypeConversionRule` from convert types other than timestamp and boolean
 - prevent optimizer from optimize plan after `MergeScan`
 - use the schema from optimized plan as the output schema of `MergeScan`
 - optimize input plan if the `MergeScan` is executed only in frontend

The error log: 
```
MySQL [(none)]>  SELECT SUM(-1) FROM numbers WHERE number=-1;
ERROR 1815 (HY000): Failed to execute query: SELECT SUM(-1) FROM numbers WHERE number=-1, source: Failed to execute logical plan, source: Failure during query execution, source: Fail to create physical plan: Internal error: The type of UInt32 Eq Int64 of binary physical should be same. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Review by evenyag — CHANGES_REQUESTED** [review]
Please fix the title.

**Review by evenyag — APPROVED** [review]
If it works not only on your machine ......

File: src/query/src/optimizer/type_conversion.rs [hunk 1]
```diff
@@ -169,12 +169,19 @@ impl TypeConverter {
...
+            return Ok((left.clone(), right.clone()));
+        }
+
         match (left, right) {
             (Expr::Column(col), Expr::Literal(value)) => {
                 let casted_right = Self::cast_scalar_value(value, left_type)?;
                 if casted_right.is_null() {
                     return Err(DataFusionError::Plan(format!(
-                        "column:{col:?} value:{value:?} is invalid",
+                        "column:{col:?}. Cast value:{value:?} to {left_type:?} is invalid",
```
Comment (v0y4g3r) [inline src/query/src/optimizer/type_conversion.rs:184]:
```suggestion
                        "column:{col:?}. Casting value:{value:?} to {left_type:?} is invalid",
```

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**